modules/core/src/main/java/org/apache/synapse/message/processor/impl/forwarder/ForwardingService.java

```diff
@@ -397,205 +442,84 @@ public void init(SynapseEnvironment se) {
         }
          return nonRetryCodes;
     }
-    
+
 	/**
-	 * Receives the next message from the message store.
-	 * 
-	 * @param msgConsumer
-	 *            message consumer
+	 * Receives the next message from the message store. On a connection issue
+	 * to store, retry will happen in this method
+	 *
 	 * @return {@link MessageContext} of the last message received from the
-	 *         store.
+	 * store.
+	 * @throws StoreForwardException on an issue fetching a message from store. Here
+	 *                               all connections to internal store will be closed. Also all connections
+	 *                               made by consumer to poll messages will be closed
 	 */
-	public MessageContext fetch(MessageConsumer msgConsumer) {
-		return messageConsumer.receive();
+	public MessageContext fetch() throws StoreForwardException {
+
+		MessageContext fetchedMessage = null;
+
+		for (int connAttempt = 0;
+			 connAttempt < maxConnectionAttemptsToStore || maxConnectionAttemptsToStore == -1;
+			 connAttempt++) {
+			try {
+				fetchedMessage = messageConsumer.receive();
+				break;
```

Comment:
Why we have this break statement? Since we have this without any condition, the first loop will break. What is the usecase here?

Reply:
fetch() method should do messageConsumer.receive() to get a message. If it is done without any exception, we break the loop and return the fetched message. If SynapseException is thrown, we need to retry in the loop until success, up to the max retries configured.